File: leetcode/java/0278. First Bad Version.java
package leetcode.java;

// Ficticious class to bypass highlighted errors
class VersionControl {
    public boolean isBadVersion(int version) {
        return true;
    }
}

// Time complexity: O(logN)
// Space complexity: O(1)

class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int left = 1;
        int right = n;

        while (right > left) {
            int mid = left + (right - left) / 2;
            if (isBadVersion(mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
}

File: leetcode/python/0278. First Bad Version.py
# Time complexity: O(logN)
# Space complexity: O(1)

def isBadVersion():
    '''
    Question has predefined this function
    '''
    pass

class Solution(object):
    def firstBadVersion(self, n):

        if isBadVersion(1):
            return 1

        left = 1
        right = n - 1

        while right >= left:
            mid = (left + right) / 2

            if isBadVersion(mid):
                right = mid - 1
            else:
                left = mid + 1
        return left